### PR TITLE
[NEW] Session endpoints 

### PR DESCRIPTION
**Pull request by albuquerquefabio**

### CODE REVIEW

File: packages/core-typings/src/ISession.ts [hunk 1]
```diff
@@ -50,3 +52,15 @@ export type UserSessionAggregation = Pick<ISession, '_id'> & {
...
 export type DeviceSessionAggregation = Pick<ISession, '_id'> & { type: string; name: string; version: string; count: number; time: number };
 export type OSSessionAggregation = Pick<ISession, '_id'> & { name: string; version: string; count: number; time: number };
+
+// Session for Multi Device Manager
+
+export interface IMDMSessionParams {
+	count?: number;
+	offset?: number;
+	search?: string | null;
+}
```
Comment (ggazzo) [inline packages/core-typings/src/ISession.ts:62]: wrong place this is not a core typing ;)

**Review by ggazzo — CHANGES_REQUESTED** [review]

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 2]
```diff
@@ -737,12 +739,333 @@ export const aggregates = {
 };
 
 export class SessionsRaw extends BaseRaw<ISession> {
+	protected indexes: IndexSpecification[] = [
+		{ key: { instanceId: 1, sessionId: 1, year: 1, month: 1, day: 1 } },
+		{ key: { instanceId: 1, sessionId: 1, userId: 1 } },
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:744]: If you change this idx to have `sessionId` as first, you won't need an individual idx on 746

Comment (KevLehman) [reply apps/meteor/app/models/server/raw/Sessions.ts:744]: Ref: https://www.mongodb.com/docs/manual/core/index-compound/#prefixes

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 3]
```diff
@@ -737,12 +739,333 @@ export const aggregates = {
...
+		},
+	];
+
 	private secondaryCollection: Collection<ISession>;
 
 	constructor(public readonly col: Collection<ISession>, public readonly colSecondary: Collection<ISession>, trash?: Collection<ISession>) {
 		super(col, trash);
 
 		this.secondaryCollection = colSecondary;
+		this.col.createIndexes(this.indexes);
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:782]: Isn't this done on `BaseRaw`? 👀

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 4]
```diff
@@ -737,12 +739,333 @@ export const aggregates = {
...
+				createdAt: {
+					$first: '$createdAt',
+				},
+				loginAt: {
+					$first: '$loginAt',
+				},
+			},
+		};
+
+		const [docTotal] = await this.col
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:871]: Instead of running the agg 2 times, you can take a look at the `$facet` stage 👀 
https://www.mongodb.com/docs/manual/reference/operator/aggregation/facet/

(If you search for $facet on the repo, you'll find 2 examples of this being used to paginate aggregations on a single call 👀 )

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 5]
```diff
@@ -737,12 +739,333 @@ export const aggregates = {
...
+					$first: '$mostImportantRole',
+				},
+				roles: {
+					$first: '$roles',
+				},
+			},
+		};
+		const limitOperator = { $limit: count };
+
+		const [docTotal] = await this.col
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:996]: Same here 😬

File: apps/meteor/app/models/server/raw/Users.js [hunk 6]
```diff
@@ -174,6 +174,31 @@ export class UsersRaw extends BaseRaw {
...
+			.project({
+				_id: 1,
+				username: 1,
+				name: 1,
+				avatarOrigin: 1,
+				avatarETag: 1,
+			})
+			.skip(offset)
+			.limit(count)
+			.toArray();
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Users.js:199]: as a note 👀 you can pass all of these calls to the 2nd param on `find`.

File: apps/meteor/ee/server/api/sessions.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+			check(
+				this.queryParams,
+				Match.ObjectIncluding({
+					offset: Match.Maybe(String),
+					count: Match.Maybe(String),
+					search: Match.Maybe(String),
+				}),
+			);
+
+			try {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:30]: no need for the try/catch if you're not doing something special with the error 👀

File: apps/meteor/ee/server/api/sessions.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+			}
+
+			if (!hasLicense('device-management')) {
+				return API.v1.unauthorized();
+			}
+			check(this.bodyParams, {
+				sessionId: String,
+			});
+
+			try {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:59]: Same here 😬

File: apps/meteor/ee/server/api/sessions.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+API.v1.addRoute(
+	'sessions/list.all',
+	{ authRequired: true, twoFactorRequired: true },
+	{
+		async get() {
+			if (!this.userId) {
+				API.v1.failure('error-invalid-user');
+			}
+			if (!hasRole(this.userId, 'admin')) {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:92]: Instead of checking for a role, can we check for a permission? 👀

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 3 above)
Comment (albuquerquefabio) [reply apps/meteor/app/models/server/raw/Sessions.ts:782]: Just works for me when I add this line. But, I need to add this before this [patch release](#25101). 
I skip it, I'll remove it!

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 4 above)
Comment (albuquerquefabio) [reply apps/meteor/app/models/server/raw/Sessions.ts:871]: So nice!
We have many things that follow this pattern above. Maybe updating this pattern is a good task that we can do in a near future. :smiley:

File: apps/meteor/ee/server/api/sessions.ts (hunk 9 above)
Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:92]: Yeah, we can do it. But in this case, we don't have this kind of permissions yet.

Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:92]: https://github.com/RocketChat/Rocket.Chat/pull/25419

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 10]
```diff
@@ -745,17 +747,313 @@ export class SessionsRaw extends BaseRaw<ISession> {
...
+		};
+
+		const queryArray = [matchOperator, sortOperator, groupOperator, projectOperator, facetOperator];
+
+		const [
+			{
+				docs: sessions,
+				count: [{ total } = { total: 0 }],
+			},
+		] = (await this.col.aggregate(queryArray).toArray()) as unknown as { docs: ISession[]; count: { total: number }[] }[];
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:878]: you can pass a type param to the aggregate function iirc

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 11]
```diff
@@ -745,17 +747,313 @@ export class SessionsRaw extends BaseRaw<ISession> {
...
+		};
+
+		const queryArray = [matchOperator, sortOperator, groupOperator, facetOperator];
+
+		const [
+			{
+				docs: sessions,
+				count: [{ total } = { total: 0 }],
+			},
+		] = (await this.col.aggregate(queryArray).toArray()) as unknown as {
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:1023]: same here 👀

File: apps/meteor/client/views/room/MessageList/hooks/useMessages.ts [hunk 12]
```diff
@@ -29,15 +31,15 @@ const removePossibleNullValues = ({
...
-	...(editedBy !== null && { editedBy }),
-	...(editedAt !== null && { editedAt }),
-	...(emoji !== null && { emoji }),
-	...(avatar !== null && { avatar }),
-	...(alias !== null && { alias }),
-	...(customFields !== null && { customFields }),
-	...(groupable !== null && { groupable }),
-	...(attachments !== null && { attachments }),
-	...(reactions !== null && { reactions }),
+	...(isNotNullOrUndefined(editedBy) && { editedBy }),
```
[the remark this answers is not in the document]
Comment (albuquerquefabio) [reply apps/meteor/client/views/room/MessageList/hooks/useMessages.ts:34]: I did not add it. Maybe some merge brings this piece of code.

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 10 above)
Comment (albuquerquefabio) [reply apps/meteor/app/models/server/raw/Sessions.ts]: Done

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 11 above)
Comment (albuquerquefabio) [reply apps/meteor/app/models/server/raw/Sessions.ts]: Done

File: packages/core-typings/src/ISession.ts [hunk 13]
```diff
@@ -27,10 +27,14 @@ export interface ISession {
 	sessionId: string;
 	_updatedAt: Date;
 	createdAt: Date;
+	loginToken?: string;
```
Comment (debdutdeb) [inline packages/core-typings/src/ISession.ts:30]:
```suggestion
```

File: apps/meteor/app/statistics/server/lib/SAUMonitor.ts [hunk 14]
```diff
@@ -190,6 +190,7 @@ export class SAUMonitorClass {
 			return;
 		}
 
+
```
Comment (debdutdeb) [inline apps/meteor/app/statistics/server/lib/SAUMonitor.ts:193]:
```suggestion
```

File: packages/core-typings/src/ISession.ts (hunk 13 above)
Comment (albuquerquefabio) [reply packages/core-typings/src/ISession.ts]: But we need this object

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 3 above)
Comment (KevLehman) [reply apps/meteor/app/models/server/raw/Sessions.ts:763]: @sampaiodiego , were you aware of this? 👀 like, index creation is not working on baseraw

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 15]
```diff
@@ -761,17 +763,315 @@ export class SessionsRaw extends BaseRaw<ISession> {
...
+
+		const [
+			{
+				docs: sessions,
+				count: [{ total } = { total: 0 }],
+			},
+		] = await this.col
+			.aggregate<{
+				docs: ISession & { _user: Pick<IUser, 'name' | 'username' | 'avatarETag' | 'avatarOrigin'> }[];
+				count: { total: number }[];
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:1042]: This part could be its own type now, something like:
`WithItemCount<T> = T & { count: { total: number }[] }`

So that you can reuse it on other places

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 16]
```diff
@@ -761,17 +763,315 @@ export class SessionsRaw extends BaseRaw<ISession> {
...
 	protected modelIndexes(): IndexSpecification[] {
 		return [
-			{ key: { instanceId: 1, sessionId: 1, year: 1, month: 1, day: 1 } },
-			{ key: { instanceId: 1, sessionId: 1, userId: 1 } },
-			{ key: { instanceId: 1, sessionId: 1 } },
-			{ key: { sessionId: 1 } },
-			{ key: { userId: 1 } },
-			{ key: { year: 1, month: 1, day: 1, type: 1 } },
-			{ key: { type: 1 } },
+			{ key: { createdAt: 1 } },
```
Comment (KevLehman) [inline apps/meteor/app/models/server/raw/Sessions.ts:1051]: A reminder that, if you removed an index that was already present on develop, pls create a migration removing the index from the db :)

File: apps/meteor/ee/server/api/sessions.ts [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+			});
+
+			const { sessionId } = this.bodyParams;
+			const sessionObj = await Sessions.findOneBySessionIdAndUserID(sessionId, this.userId);
+
+			if (!sessionObj) {
+				return API.v1.notFound('Session not found');
+			}
+
+			Meteor.users.update(this.userId, {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:62]: Can the "normal" user model be used here? 👀

File: apps/meteor/ee/server/api/sessions.ts [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+
+API.v1.addRoute(
+	'sessions/list.all',
+	{ authRequired: true, twoFactorRequired: true },
+	{
+		async get() {
+			if (!this.userId) {
+				API.v1.failure('error-invalid-user');
+			}
+			if (!hasPermission(this.userId, 'view-device-management')) {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:84]: You can move this check to the 2 param of `addRoute` as `permissionsRequired: ['view-device-management']`

File: apps/meteor/ee/server/api/sessions.ts [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+			}
+			if (!hasLicense('device-management')) {
+				return API.v1.unauthorized();
+			}
+
+			check(this.bodyParams, {
+				sessionId: String,
+			});
+
+			try {
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:139]: No need to use try/catch on handler if you're not doing something special with the error value 👀

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 3 above)
Comment (pierre-lehnen-rc) [reply apps/meteor/app/models/server/raw/Sessions.ts:763]: It was fixed one or two releases ago.

File: apps/meteor/ee/server/api/sessions.ts (hunk 17 above)
Comment (pierre-lehnen-rc) [reply apps/meteor/ee/server/api/sessions.ts:62]: Yes, `Meteor.users` shouldn't be used.
And preferably the raw model should be used instead of the normal one.

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 20]
```diff
@@ -1295,17 +1599,22 @@ export class SessionsRaw extends BaseRaw<ISession> {
...
+				logoutBy: userId,
 			},
 		};
 
-		return this.updateMany(query, update);
+		if (!session?.loginToken) {
+			return this.updateOne(query, update);
+		}
+
+		return this.updateMany({ userId, loginToken: session?.loginToken }, update);
```
Comment (pierre-lehnen-rc) [inline apps/meteor/app/models/server/raw/Sessions.ts:1617]: Why was this change necessary?

I'm not following the logic here.

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 15 above)
Comment (pierre-lehnen-rc) [reply apps/meteor/app/models/server/raw/Sessions.ts:1042]: I agree. And since not all session attributes are being projected, the resulting type is not a full `ISession`, in this case it's better to set the type picking only the attributes that are actually projected.

File: packages/core-typings/src/ISession.ts [hunk 21]
```diff
@@ -30,7 +30,10 @@ export interface ISession {
 	host: string;
 	ip: string;
 	loginAt: Date;
+	logoutAt?: Date;
 	closedAt?: Date;
+	logoutBy?: string;
```
Comment (pierre-lehnen-rc) [inline packages/core-typings/src/ISession.ts:35]:
```suggestion
	logoutBy?: IUser['_id'];
```
When the attribute name doesn't specify what kind of information is in the attribute, it's a good practice to link the type directly like this so we don't have to look up if this uses an id or an username or something else.

**Review by pierre-lehnen-rc — CHANGES_REQUESTED** [review]

File: apps/meteor/app/models/server/raw/Sessions.ts (hunk 3 above)
Comment (KevLehman) [reply apps/meteor/app/models/server/raw/Sessions.ts:763]: I quite remember that, but was not sure, thanks Pierre!

So, @albuquerquefabio , you can trust on BaseRaw to create indexes now 👀

File: apps/meteor/ee/server/api/sessions.ts [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+
+				Meteor.users.update(sessionObj.userId, {
+					$pull: {
+						'services.resume.loginTokens': {
+							hashedToken: sessionObj.loginToken,
+						},
+					},
+				});
+
+				await Sessions.updateMany({ loginToken: sessionObj.loginToken }, { $set: { logoutAt: new Date(), logoutBy: this.userId } });
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:155]: Pls move this update query to the model 👀

File: apps/meteor/ee/server/api/sessions.ts [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+			const sortKeys = ['loginAt', 'device.name', 'device.os.name', '_user.username', '_user.name'];
+			if (!Object.keys(sort).filter((key) => sortKeys.includes(key)).length) {
+				return API.v1.failure('error-invalid-sort');
+			}
+
+			const searchUser = await Users.find<Pick<IUser, '_id'>>(
+				{ $text: { $search: search }, active: true },
+				{
+					projection: { _id: 1, score: { $meta: 'textScore' } },
```
[the remark this answers is not in the document]
Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:125]: No, it is to sort by the best text score in this case. I need only the Ids.

File: apps/meteor/ee/server/api/sessions.ts [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+import { IUser } from '@rocket.chat/core-typings';
+
+import { Users, Sessions } from '../../../app/models/server/raw/index';
+import { API } from '../../../app/api/server/api';
+import { hasLicense } from '../../app/license/server/license';
+
+API.v1.addRoute(
+	'sessions/list',
+	{ authRequired: true },
+	{
```
[the remark this answers is not in the document]
Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:11]: Yes, but an EE key is necessary. This endpoint only shows information about the user who requested it.

File: apps/meteor/app/models/server/raw/Sessions.ts [hunk 25]
```diff
@@ -30,6 +33,9 @@ type DestructuredDateWithType = {
 type DestructuredRange = { start: DestructuredDate; end: DestructuredDate };
 type DateRange = { start: Date; end: Date };
 
+type CustomSortOp = 'loginAt' | 'device.name' | 'device.os.name';
+type CustomSortOpAdmin = CustomSortOp | 'loginAt' | 'device.name' | 'device.os.name' | '_user.username' | '_user.name';
```
[the remark this answers is not in the document]
Comment (albuquerquefabio) [reply apps/meteor/app/models/server/raw/Sessions.ts:37]: Thank you!

File: apps/meteor/ee/server/api/sessions.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:125]: I made some tests here and regarding the score on projection is optional. We can use the sort to organize the best content based on the search score.

File: apps/meteor/ee/definition/rest/v1/sessions/sessions.ts [hunk 26]
```diff
@@ -0,0 +1,29 @@
+import type { ISession, DeviceManagementSession, DeviceManagementPopulatedSession } from '@rocket.chat/core-typings';
+
+import type { PaginatedRequest } from '../../../../../../../packages/rest-typings/src/helpers/PaginatedRequest';
```
Comment (KevLehman) [inline apps/meteor/ee/definition/rest/v1/sessions/sessions.ts:3]: Can't you import these form `@rocket-chat/rest-typings`? 👀

File: apps/meteor/ee/definition/rest/v1/sessions/sessions.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+import type { PaginatedResult } from '../../../../../../../packages/rest-typings/src/helpers/PaginatedResult';
+import { SessionsProps } from './SessionsProps';
```
Comment (KevLehman) [inline apps/meteor/ee/definition/rest/v1/sessions/sessions.ts:5]: Do `import type` work here? 👀

File: apps/meteor/ee/server/api/sessions.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/ee/server/api/sessions.ts:12]: Nice!

File: apps/meteor/ee/server/api/sessions.ts [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+import { IUser } from '@rocket.chat/core-typings';
+
+import { isSessionsProps } from '../../definition/rest/v1/sessions/SessionsProps';
+import { Users, Sessions } from '../../../app/models/server/raw/index';
+import { API } from '../../../app/api/server/api';
+import { hasLicense } from '../../app/license/server/license';
+
+API.v1.addRoute(
+	'sessions/list',
+	{ authRequired: true },
```
Comment (KevLehman) [inline apps/meteor/ee/server/api/sessions.ts:11]: This is the only handler here using Check instead of AJV 🙈

Comment (albuquerquefabio) [reply apps/meteor/ee/server/api/sessions.ts:11]: No issues

**Review by KevLehman — APPROVED** [review]
All good for me :) Let's see other reviewer's comments
